### PR TITLE
user-cluster tool

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:
Syncier asked for CI tool to automate creation of user-clusters

```release-note
user-cluster tool, to automate user-cluster creation
```

### CODE REVIEW

File: api/cmd/user-cluster/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+### Example AWS Cluster spec
+
+    apiVersion: kubermatic.k8s.io/v1
+    kind: Cluster
+    metadata:
+      name: aws-cicd-tmp-cluster
+      labels:
+        user: << Optional UserID to see cluster in UI>>
+        worker-name: ""
+    spec:
```
Comment (thetechnick) [inline api/cmd/user-cluster/README.md:32]: This does not look right

File: api/cmd/user-cluster/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      aws:
+        accessKeyId: << your AWS accessKeyId >>
+        secretAccessKey: << your AWS secretAccessKey >>
+    humanReadableName: aws-cicd-tmp-test-runner
+    pause: false
+    version: 1.10.5
+
+### Example AWS Node spec
+
+    spec:
```
Comment (thetechnick) [inline api/cmd/user-cluster/README.md:44]: can you please add the rest of the object?

File: api/cmd/user-cluster/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        kubermatic/kubermatic user-cluster
+
+## Flags
+TBD
+
+## Examples
+
+### Example AWS Cluster spec
+
+    apiVersion: kubermatic.k8s.io/v1
```
Comment (thetechnick) [inline api/cmd/user-cluster/README.md:25]: Can you please highlight both code blocks as yaml?

File: api/cmd/user-cluster/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    $ docker run --rm -it \
+        -w /app \
+        -e KUBECONFIG=/config/kubeconfig \
+        -v /host_path/to/kubeconfig:/config/kubeconfig \
+        -v /host_path/to/cluster_template.yaml:/app/cluster.yaml \
+        -v /host_path/to/node_template.yaml:/app/node.yaml \
+        kubermatic/kubermatic user-cluster
+
+## Flags
+TBD
```
Comment (thetechnick) [inline api/cmd/user-cluster/README.md:19]: Please document the current flags or remove this section

File: api/cmd/user-cluster/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# Automate creation of kubermatic managed clusters
+
+## What is this
+It's a tool to automate creating of custom resources for kubermatic managed
```
Comment (thetechnick) [inline api/cmd/user-cluster/README.md:4]: remove line break

File: api/cmd/user-cluster/cluster_creator.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+	apiv1 "github.com/kubermatic/kubermatic/api/pkg/api/v1"
+	kubermaticclientset "github.com/kubermatic/kubermatic/api/pkg/crd/client/clientset/versioned"
+	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+	"github.com/kubermatic/kubermatic/api/pkg/provider"
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	machineresource "github.com/kubermatic/kubermatic/api/pkg/resources/machine"
+)
+
+const (
+	generateName = "auto-cluster"
```
Comment (thetechnick) [inline api/cmd/user-cluster/cluster_creator.go:35]: the user should be able to configure this by himself (new flag?)

File: api/cmd/user-cluster/cluster_creator.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+	}
+
+	machines := clusterClient.ClusterV1alpha1().Machines(metav1.NamespaceSystem)
+	template, err := machineresource.Machine(cluster, &node, dc, nil)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < ctl.runOpts.Nodes; i++ {
+		template.Name = fmt.Sprintf("%s%s", generateName, rand.String(5))
```
Comment (thetechnick) [inline api/cmd/user-cluster/cluster_creator.go:234]: Use `template.GenerateName` as a base when generating the machine name

**Review by thetechnick — CHANGES_REQUESTED** [review]

File: api/cmd/user-cluster/README.md (hunk 1 above)
Comment (kron4eg) [reply api/cmd/user-cluster/README.md:32]: please elaborate what do you mean

File: api/cmd/user-cluster/README.md (hunk 2 above)
Comment (kron4eg) [reply api/cmd/user-cluster/README.md:44]: it's complete object (used by kubermatic http api) github.com/kubermatic/kubermatic/api/pkg/api/v1.Node

File: api/cmd/user-cluster/README.md (hunk 5 above)
Comment (kron4eg) [reply api/cmd/user-cluster/README.md:4]: why? it's 80-th column text-wrap

**Review by alvaroaleman — CHANGES_REQUESTED** [review]
Why do we put a project-specific tool into the main kubermatic/kubermatic repo?

And why don't we leverage our API?

File: api/cmd/user-cluster/cluster_creator.go (hunk 6 above)
Comment (kron4eg) [reply api/cmd/user-cluster/cluster_creator.go:35]: I don't see a reason why. That tool is for CI. It creates those "automatic user clusters".

**Comment by kron4eg** [comment]
@alvaroaleman 
>Why do we put a project-specific tool into the main kubermatic/kubermatic repo?

First of all, although `user-cluster` is for syncier, it will be useful for us too (for CI purpose), and for everyone basically. Second, since kubermatic/kubermatic is a monorepo where else should we put this tool?

>And why don't we leverage our API?

That'd mean to start whole new project, generating api-client out of swagger spec, wrapping everything in nice CLI, dealing with (non-existing yet Service Accounts). That's not an objective for this tool. What we need here is to:

* create user-cluster control-plain
* create machines
* dump user-cluster kubeconfig

**Comment by alvaroaleman** [comment]
> First of all, although user-cluster is for syncier, it will be useful for us too (for CI purpose), and for everyone basically. Second, since kubermatic/kubermatic is a monorepo where else should we put this tool?

We already have the testrunner for that and don't need another tool

> That'd mean to start whole new project, generating api-client out of swagger spec, wrapping everything in nice CLI, dealing with (non-existing yet Service Accounts). That's not an objective for this tool. What we need here is to:

Is there a time constraint on this? UI folks apparently already found a way to get tokens for usage in scripts in their testing. It should be pretty easy after that

**Comment by kron4eg** [comment]
>We already have the testrunner

Please point me to one

>Is there a time constraint on this?

As always, like, previous week :)

**Comment by alvaroaleman** [comment]
> Please point me to one

https://github.com/kubermatic/kubermatic/tree/c5853e722f9b47fca019759021f38470cd7aad7c/api/cmd/conformance-tests

>> Is there a time constraint on this?

> As always, like, previous week :)

Can you still check in with the api team how difficult that would be? I don't think this will take an incredible amount of time

**Comment by kron4eg** [comment]
@alvaroaleman those are incomparable. `conformance-tests` is for running "conformance-tests" statically compiled into it...
`user-cluster` is general purpose tool that will only prepare user-cluster and let users do whatever they want to do with a cluster.

**Comment by alvaroaleman** [comment]
> @alvaroaleman those are incomparable. conformance-tests is for running "conformance-tests" statically compiled into it...

Yes I know they are not the same but for our tests which you stated an example on which we ourselves need the code you add here we do not need this code because we already have the `conformance-tests` cli

> user-cluster is general purpose tool that will only prepare user-cluster and let users do whatever they want to do with a cluster.

We specifically do not want a general-purpose tool that does not use the api

**Comment by kron4eg** [comment]
@alvaroaleman I'm not sure I follow where this leads

>alvaroaleman requested changes an hour ago

so which changes are you requesting exactly?

**Comment by alvaroaleman** [comment]
> so which changes are you requesting exactly?

This tool to use our API instead of directly working on CRDs

**Comment by kron4eg** [comment]
what's the problem to work directly on CRDs?

**Comment by alvaroaleman** [comment]
> what's the problem to work directly on CRDs?

The crds are an implementation detail of ours and not an API we want to expose. We want and do change them without notifying external parties

**Comment by bashofmann** [comment]
> The crds are an implementation detail of ours and not an API we want to expose. We want and do change them without notifying external parties

I would argue that CRDs are a public API and should not be changed without notifying external parties, and if they are changed backwards compatibility should be kept in mind and CRDs should be versioned correctly.

**Comment by alvaroaleman** [comment]
> I would argue that CRDs are a public API and should not be changed without notifying external parties, and if they are changed backwards compatibility should be kept in mind and CRDs should be versioned correctly.

No, neither kubermatic cluster crd nor the machine crd are public apis. The first one is only used internally and should not be used externally, the second one is declared alpha by upstream which means they do incompatible changes to it and the pace at which we do or do not adapt upstream changes is a call we make.

CRD versioning is currently not possible, it got [merged two days ago into kubernetes master](https://github.com/kubernetes/kubernetes/pull/67006) and will _maybe_ land in 1.13. Right now it is possible to specify more than one version but all of them must be identical because there is no conversion.
Even if it was possible with 1.13 we won't use it for the cluster-api types, as we support Kube back to 1.10 for user clusters which does not have crd versioning.

We may decide to pick up crd versioning for Kubermatic but even then, the kubermatic crd is not a public api. Our API is the public api.

**Comment by kron4eg** [comment]
>Our API is the public api.

And we don't have anything to work with it besides swagger file, right?

**Comment by alvaroaleman** [comment]
> And we don't have anything to work with it besides swagger file, right?

Yes. You can generate a client from the swagger thought, @glower already did that at some point

**Comment by toschneck** [comment]
@kron4eg, @alvaroaleman my thoughts on this:
* For the most people CRDs are seen as public APIs. If this is correct (regarding to versioning issues) or not is another thing, but if people will see CRDs on the cluster, they will use it. Also popular frameworks like Prometheus use CRDs as abstraction layer for there API, [coreos/prometheus-operator#customresourcedefinitions](https://github.com/coreos/prometheus-operator#customresourcedefinitions). We also have the machine CRDs in the user cluster whats part of the ClusterAPI, so for me this an abstraction layer for the real API of the operator.

* The purpose of the tool is, that we run CI/CD load of customer in automated provisioned "ad-hoc  clusters". So yeah it would be better to keep it outside of main project and just import the types. But it don't know how you guys manage your dependencies.  In my opinion this tool need to be simple and flexible. The basic idea was following use case:
1. CI job trigger the user-cluster-tool
2. tool create a cluster based on some `cluster.yaml` CRD template
3. extract the kubeconfig for the new created cluster
3. run the CI/CD stuff with the applied kubeconfig
4. CI job call tool again  and cleanup the cluster.

To reach this, using the CRDs is for me the easiest and most flexible way. If the CRDs are not part of an API the whole clusterAPI project sounds weird for me.

**Pull request closed without merging**